### PR TITLE
added helper function that joins vectors of strings

### LINKED ISSUE #4640: Add helper function that joins vectors of strings
Add a helper function to `CommonData.h` that takes a vector of strings (or perhaps just a template that can be used in a range-based for loop) that does the following:

Returns a single string where all elements are joined by a separator (default is `", "`) where optionally the last separator can be different (for example `" or "`).

The first one can be solved using boost::algorithm::join, but no the second use-case.

Example:

`["one", "to", "three"] -> `"one, to or three"`

### PR DESCRIPTION
**Pull request by evgeniuz**
### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] New tests have been created which fail without the change (if possible)
- [x] README / documentation was extended, if necessary
- [x] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

### Description
Added helper function that joins vectors of strings and tests for it.

Fixes #4640.

### CODE REVIEW

**Comment by axic** [comment]
This looks great, thank you!

Can you also use this new helper in some places? One place using it is the "did you mean?" suggestion messages for unknown variable names.

File: libdevcore/CommonData.h [hunk 1]
```diff
@@ -47,6 +49,32 @@ enum class HexPrefix
...
+template<class T>
+std::string join(
+	const T& _list,
+	const std::string& _separator =", ",
+	const std::string& _lastSeparator = ""
+) {
+	std::string result = boost::algorithm::join(_list, _separator);
+
+	// calling .back() on empty _list is undefined behavior
+	if (!_lastSeparator.empty() && !_list.empty()) {
```
Comment (axic) [inline libdevcore/CommonData.h]: Should this allow empty lists as input?

If yes, should we just have a `if (_list.empty()) return string{}` as the first statement before calling `join`?

**Review by chriseth — CHANGES_REQUESTED** [review]
Thanks for the pull request!

Could you please update the coding style? We generally use `std::string const& _separator`, braces on a line of their own and tabs for indentation.

Also I think that the case where there is a last separator looks a little hacking. Perhaps it is best to just implement everything manually?

**Comment by evgeniuz** [comment]
Updated coding style, refactored function in a less hackish way (it's actually better that way, now it can work with arrays as well, as it doesn't depend on `_list.back()` and `_list.empty()` anymore), also refactored `quotedAlternativesList` to use this new function.

File: libdevcore/CommonData.h [hunk 2]
```diff
@@ -56,20 +54,34 @@ enum class HexPrefix
...
-	std::string result = boost::algorithm::join(_list, _separator);
-
-	// calling .back() on empty _list is undefined behavior
-	if (!_lastSeparator.empty() && !_list.empty()) {
-		boost::replace_last(
-			result,
-			_separator + _list.back(),
-			_lastSeparator + _list.back()
-		);
+std::string join
```
Comment (axic) [inline libdevcore/CommonData.h]: I think this should be called something like `joinHumanReadablelList`. Any better suggestions?

Comment (axic) [reply libdevcore/CommonData.h]: Also likely it should go into the `StringUtils` file.

Comment (evgeniuz) [reply libdevcore/CommonData.h]: ok, will rename and move it right now.

Comment (evgeniuz) [reply libdevcore/CommonData.h]: maybe just joinHumanReadable?

File: libdevcore/StringUtils.h [hunk 3]
```diff
@@ -36,4 +36,43 @@ size_t stringDistance(std::string const& _str1, std::string const& _str2);
...
+/// Joins collection of strings into one string with separators between, last separator can be different.
+/// @param _list collection of strings to join
+/// @param _separator defaults to ", "
+/// @param _lastSeparator (optional) will be used to separate last two strings instead of _separator
+/// @example join(vector<string>{"a", "b", "c"}, "; ", " or ") == "a; b or c"
+template<class T>
+std::string joinHumanReadable
+(
+	T const& _list,
+	std::string const& _separator =", ",
```
Comment (chriseth) [inline libdevcore/StringUtils.h]: Please add a space after `=`

File: libdevcore/StringUtils.h [hunk 4]
```diff
@@ -36,4 +36,43 @@ size_t stringDistance(std::string const& _str1, std::string const& _str2);
...
+
+	std::string result;
+
+	// append first string
+	if (it != itEnd) {
+		result += *it;
+		++it;
+	}
+
+	for (;it != itEnd; ++it) {
```
Comment (chriseth) [inline libdevcore/StringUtils.h]: Please put the `{` on a line of its own just as with function bodies (also above)

File: libdevcore/StringUtils.h [hunk 5]
```diff
@@ -36,4 +36,43 @@ size_t stringDistance(std::string const& _str1, std::string const& _str2);
...
+		if ((next(it) == itEnd) && !_lastSeparator.empty()) {
+			// last iteration
+			result += _lastSeparator;
+		} else {
```
Comment (chriseth) [inline libdevcore/StringUtils.h]: Please omit braces if there is only a single statement.

**Review by chriseth — CHANGES_REQUESTED** [review]
Wonderful! There are only some small style changes left.

**Comment by axic** [comment]
@evgeniuz as a last step when all the changes are agreed on, can you please squash the commits down to a single one?

You could use `git rebase -i HEAD~4` to do it.

**Comment by axic** [comment]
@chriseth do we have any more comments? I am not sure about the function name, but have no good suggestions.

**Review by axic — APPROVED** [review]

**Pull request merged**